modules/rtdModule/index.js

```diff
@@ -0,0 +1,208 @@
+import * as utils from '../../src/utils';
+
+/** @type {string} */
+const MODULE_NAME = 'realTimeData';
+/** @type {RtdSubmodule[]} */
+let subModules = [];
+/** @type {ModuleConfig} */
+let _moduleConfig;
+
+/**
+ * enable submodule in User ID
+ * @param {RtdSubmodule} submodule
+ */
+export function attachRealTimeDataProvider(submodule) {
+  subModules.push(submodule);
+}
+
+export function init(config) {
+  const confListener = config.getConfig(MODULE_NAME, ({realTimeData}) => {
+    if (!realTimeData.dataProviders || typeof (realTimeData.auctionDelay) == 'undefined') {
+      utils.logError('missing parameters for real time module');
+      return;
+    }
+    confListener(); // unsubscribe config listener
+    _moduleConfig = realTimeData;
+    // delay bidding process only if auctionDelay > 0
+    if (!_moduleConfig.auctionDelay > 0) {
+      getHook('bidsBackCallback').before(setTargetsAfterRequestBids);
+    } else {
+      getGlobal().requestBids.before(requestBidsHook);
+    }
+  });
+}
+
+/**
+ * get data from sub module
+ * @param {AdUnit[]} adUnits received from auction
+ * @param {function} callback callback function on data received
+ */
+function getProviderData(adUnits, callback) {
+  const callbackExpected = subModules.length;
+  let dataReceived = [];
+  let processDone = false;
+  const dataWaitTimeout = setTimeout(() => {
+    processDone = true;
+    callback(dataReceived);
+  }, _moduleConfig.auctionDelay);
+
+  subModules.forEach(sm => {
+    sm.getData(adUnits, onDataReceived);
+  });
+
+  function onDataReceived(data) {
+    if (processDone) {
+      return
+    }
+    dataReceived.push(data);
+    if (dataReceived.length === callbackExpected) {
+      processDone = true;
+      clearTimeout(dataWaitTimeout);
+      callback(dataReceived);
+    }
+  }
+}
+
+/**
+ * run hook after bids request and before callback
+ * get data from provider and set key values to primary ad server
+ * @param {function} next - next hook function
+ * @param {AdUnit[]} adUnits received from auction
+ */
+export function setTargetsAfterRequestBids(next, adUnits) {
+  getProviderData(adUnits, (data) => {
+    if (data && Object.keys(data).length) {
+      const _mergedData = deepMerge(data);
+      if (Object.keys(_mergedData).length) {
+        setDataForPrimaryAdServer(_mergedData);
+      }
+    }
+    next(adUnits);
+  });
+}
+
+/**
+ * deep merge array of objects
+ * @param {array} arr - objects array
+ * @return {Object} merged object
+ */
+export function deepMerge(arr) {
+  if (!arr.length) {
+    return {};
+  }
+  return arr.reduce((merged, obj) => {
+    for (let key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        if (!merged.hasOwnProperty(key)) merged[key] = obj[key];
+        else {
+          // duplicate key - merge values
+          const dp = obj[key];
+          for (let dk in dp) {
+            if (dp.hasOwnProperty(dk)) merged[key][dk] = dp[dk];
+          }
+        }
+      }
+    }
+    return merged;
+  }, {});
```

Comment:
What is the reasoning for not just doing:
```javascript
  return arr.reduce((merged, obj) => {
    return Object.assign(merged, obj);
  }, {});
```

Am I missing a use case where this would not do what is intended by the `deepMerge`?

Reply:
Object.assign is not dealing well with duplicate keys.
See an example here - 
https://codepen.io/omerBrowsi/pen/oNNoLgp?editors=1112

Reply:
Ahh I see what you mean on the use case now.